File: swarmx/core/agent.py
"""
SwarmX Agent — Base agent class with event-driven lifecycle.

Adapted from OpenClaw's agent architecture: agents have isolated state,
bind to a provider declaratively, and communicate exclusively through
the event bus. No direct agent-to-agent calls are permitted.

The agent lifecycle follows:
  1. initialize() — one-time setup, subscribe to events
  2. on_event()   — handle incoming events
  3. think()      — process messages through the provider
  4. emit()       — publish results back to the bus
  5. shutdown()   — cleanup
"""

from __future__ import annotations

import asyncio
import logging
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

from swarmx.core.event_bus import Event, EventBus, EventPriority
from swarmx.core.provider import (
    CompletionResponse,
    Message,
    ProviderBase,
    ProviderRegistry,
    Role,
)

logger = logging.getLogger("swarmx.agent")


class AgentState(Enum):
    """Lifecycle states for an agent."""

    CREATED = "created"
    INITIALIZING = "initializing"
    IDLE = "idle"
    PROCESSING = "processing"
    ERROR = "error"
    SHUTDOWN = "shutdown"


@dataclass
class AgentConfig:
    """
    Declarative agent configuration.

    Agents are defined by config — the provider, subscriptions,
    system prompt, and tools are all declared upfront. This enables
    config-driven swarm definitions via YAML.
    """

    name: str
    provider: str  # Provider name from the registry
    model: str = ""  # Override provider's default model
    system_prompt: str = ""
    subscriptions: list[str] = field(default_factory=list)
    tools: list[dict[str, Any]] = field(default_factory=list)
    max_history: int = 50
    temperature: float | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


class Agent:
    """
    Base agent class for SwarmX.

    Agents are autonomous units that:
      - Subscribe to events via the EventBus
      - Process events through an LLM provider
      - Emit response events back to the bus

    Subclass this to create specialized agents with custom
    event handling, tool execution, or output formatting.
    """

    def __init__(
        self,
        config: AgentConfig,
        event_bus: EventBus,
        provider_registry: ProviderRegistry,
    ) -> None:
        self.config = config
        self.event_bus = event_bus
        self.provider_registry = provider_registry

        self.agent_id: str = f"{config.name}-{uuid.uuid4().hex[:6]}"
        self.state: AgentState = AgentState.CREATED
        self._message_history: list[Message] = []
        self._provider: ProviderBase | None = None
        self._processing_lock = asyncio.Lock()

    # ── Lifecycle ───────────────────────────────────────────────────

    async def initialize(self) -> None:
        """
        Initialize the agent: resolve provider, set up subscriptions.

        Called once by the engine before the agent starts processing.
        """
        self.state = AgentState.INITIALIZING

        # Resolve provider from registry
        self._provider = self.provider_registry.get(self.config.provider)
        logger.info(
            "Agent '%s' bound to provider '%s' (%s)",
            self.agent_id,
            self.config.provider,
            self._provider,
        )

        # Set system prompt
        if self.config.system_prompt:
            self._message_history.append(
                Message(role=Role.SYSTEM, content=self.config.system_prompt)
            )

        # Subscribe to configured topics
        for topic in self.config.subscriptions:
            self.event_bus.subscribe(
                topic=topic,
                handler=self._handle_event,
                subscriber_id=self.agent_id,
            )
            logger.debug("Agent '%s' subscribed to '%s'", self.agent_id, topic)

        self.state = AgentState.IDLE
        logger.info("Agent '%s' initialized and ready", self.agent_id)

    async def shutdown(self) -> None:
        """Gracefully shut down the agent."""
        self.state = AgentState.SHUTDOWN
        self.event_bus.unsubscribe(self.agent_id)
        logger.info("Agent '%s' shut down", self.agent_id)

    # ── Event Handling ──────────────────────────────────────────────

    async def _handle_event(self, event: Event) -> None:
        """Internal event handler — delegates to on_event with locking."""
        async with self._processing_lock:
            self.state = AgentState.PROCESSING
            try:
                await self.on_event(event)
            except Exception:
                self.state = AgentState.ERROR
                logger.exception("Agent '%s' error processing event %s", self.agent_id, event.topic)
                await self.emit(
                    "agent.error",
                    {
                        "agent_id": self.agent_id,
                        "event_topic": event.topic,
                        "error": "Processing failed",
                    },
                )
            finally:
                if self.state == AgentState.PROCESSING:
                    self.state = AgentState.IDLE

    async def on_event(self, event: Event) -> None:
        """
        Handle an incoming event.

        Default implementation extracts a user message from the event
        payload and sends it through the provider. Override this for
        custom event handling logic.
        """
        # Extract the message content from the event
        content = event.payload.get("content", "")
        if not content:
            content = event.payload.get("message", "")
        if not content:
            logger.debug("Agent '%s' ignoring event with no content: %s", self.agent_id, event.topic)
            return

        # Think and respond
        response = await self.think(content)
        if response:
            await self.emit(
                f"agent.response.{self.config.name}",
                {
                    "agent_id": self.agent_id,
                    "content": response.message.content,
                    "model": response.model,
                    "usage": response.usage,
                    "source_event": event.event_id,
                },
            )

    # ── Core Intelligence ───────────────────────────────────────────

    async def think(self, user_input: str) -> CompletionResponse | None:
        """
        Process user input through the LLM provider.

        Manages conversation history and returns the provider's response.
        """
        if not self._provider:
            logger.error("Agent '%s' has no provider bound", self.agent_id)
            return None

        # Add user message to history
        self._message_history.append(Message(role=Role.USER, content=user_input))

        # Trim history if needed
        if len(self._message_history) > self.config.max_history:
            # Keep system prompt + recent messages
            system_msgs = [m for m in self._message_history if m.role == Role.SYSTEM]
            recent = self._message_history[-(self.config.max_history - len(system_msgs)):]
            self._message_history = system_msgs + recent

        # Call provider
        kwargs: dict[str, Any] = {}
        if self.config.model:
            kwargs["model"] = self.config.model
        if self.config.temperature is not None:
            kwargs["temperature"] = self.config.temperature

        response = await self._provider.complete(
            messages=self._message_history,
            tools=self.config.tools or None,
            **kwargs,
        )

        # Add assistant response to history
        self._message_history.append(response.message)

        logger.debug(
            "Agent '%s' completed thinking (tokens: %s)",
            self.agent_id,
            response.usage,
        )
        return response

    # ── Event Emission ──────────────────────────────────────────────

    async def emit(
        self,
        topic: str,
        payload: dict[str, Any],
        priority: EventPriority = EventPriority.NORMAL,
    ) -> None:
        """Publish an event to the bus from this agent."""
        event = Event(
            topic=topic,
            payload=payload,
            source=self.agent_id,
            priority=priority,
        )
        await self.event_bus.publish(event)

    # ── Introspection ───────────────────────────────────────────────

    @property
    def history(self) -> list[Message]:
        """Current message history."""
        return list(self._message_history)

    @property
    def provider(self) -> ProviderBase | None:
        """The agent's bound provider."""
        return self._provider

    def __repr__(self) -> str:
        return (
            f"<Agent '{self.agent_id}' state={self.state.value} "
            f"provider={self.config.provider}>"
        )

File: src/core/agent.ts
/**
 * SwarmX Agent — Production-grade agent with full lifecycle.
 *
 * Features:
 *   - Event-driven communication via the EventBus
 *   - Session persistence (survives restarts)
 *   - Context window management (auto-prune / compact)
 *   - Tool execution loop
 *   - Usage tracking per agent
 */

import { randomUUID } from "node:crypto";
import { EventBus, createEvent, type SwarmEvent, EventPriority } from "./event-bus.js";
import {
    type ProviderBase,
    type ProviderRegistry,
    type Message,
    type CompletionResponse,
    type ToolDefinition,
    Role,
} from "./provider.js";
import { ToolExecutor, type ToolFunction } from "./tool-executor.js";
import { ContextManager, type ContextConfig } from "./context.js";
import { SessionStore } from "./session.js";
import { UsageTracker } from "./usage.js";
import { createLogger } from "../utils/logger.js";

const log = createLogger("Agent");

export enum AgentState {
    CREATED = "created",
    INITIALIZING = "initializing",
    IDLE = "idle",
    PROCESSING = "processing",
    ERROR = "error",
    SHUTDOWN = "shutdown",
}

export interface AgentConfig {
    name: string;
    provider: string;
    model?: string;
    systemPrompt?: string;
    subscriptions?: string[];
    tools?: ToolDefinition[];
    maxHistory?: number;
    temperature?: number;
    metadata?: Record<string, unknown>;
    context?: ContextConfig;
    persistSessions?: boolean;
}

export class Agent {
    readonly agentId: string;
    state: AgentState = AgentState.CREATED;

    protected messageHistory: Message[] = [];
    protected provider: ProviderBase | null = null;
    private processingLock = false;
    private log: ReturnType<typeof createLogger>;

    // Production features
    readonly toolExecutor: ToolExecutor;
    readonly contextManager: ContextManager;
    readonly usageTracker: UsageTracker;
    private sessionStore: SessionStore | null = null;
    private currentSessionId: string | null = null;

    constructor(
        readonly config: AgentConfig,
        protected readonly eventBus: EventBus,
        protected readonly providerRegistry: ProviderRegistry,
        sessionStore?: SessionStore,
    ) {
        this.agentId = `${config.name}-${randomUUID().slice(0, 6)}`;
        this.log = createLogger(`Agent:${this.agentId}`);

        this.toolExecutor = new ToolExecutor({ maxIterations: 10, toolTimeoutMs: 30_000 });
        this.contextManager = new ContextManager(config.context);
        this.usageTracker = new UsageTracker();

        if (sessionStore || config.persistSessions !== false) {
            this.sessionStore = sessionStore ?? new SessionStore();
        }
    }

    // ── Lifecycle ───────────────────────────────────────────────

    async initialize(): Promise<void> {
        this.state = AgentState.INITIALIZING;

        this.provider = this.providerRegistry.get(this.config.provider);
        this.log.info(`Bound to provider: ${this.config.provider}`);

        // Load or create session
        if (this.sessionStore) {
            const session = this.sessionStore.getOrCreate(this.agentId);
            this.currentSessionId = session.sessionId;
            if (session.messages.length > 0) {
                this.messageHistory = [...session.messages];
                this.log.info(`Restored session: ${session.messages.length} messages`);
            }
        }

        // Add system prompt if not already present
        if (this.config.systemPrompt && !this.messageHistory.some((m) => m.role === Role.SYSTEM)) {
            this.messageHistory.unshift({ role: Role.SYSTEM, content: this.config.systemPrompt });
        }

        // Subscribe to events
        const topics = this.config.subscriptions ?? [];
        for (const topic of topics) {
            this.eventBus.subscribe(topic, (event) => this.handleEvent(event), this.agentId);
        }

        this.state = AgentState.IDLE;
        this.log.info(`Ready (subscriptions: ${topics.join(", ") || "none"}, tools: ${this.toolExecutor.registeredTools.length})`);
    }

    async shutdown(): Promise<void> {
        // Persist session before shutdown
        if (this.sessionStore && this.currentSessionId) {
            this.sessionStore.save(this.currentSessionId);
            this.log.debug("Session saved");
        }

        this.state = AgentState.SHUTDOWN;
        this.eventBus.unsubscribe(this.agentId);
        this.log.info(`Shut down (usage: ${this.usageTracker.summary.costFormatted})`);
    }

    // ── Tools ───────────────────────────────────────────────────

    /**
     * Register a tool the agent can use during think().
     */
    registerTool(
        name: string,
        description: string,
        parameters: Record<string, unknown>,
        fn: ToolFunction,
    ): void {
        this.toolExecutor.register(name, description, parameters, fn);
    }

    // ── Event Handling ──────────────────────────────────────────

    private async handleEvent(event: SwarmEvent): Promise<void> {
        if (this.processingLock) return;
        this.processingLock = true;
        this.state = AgentState.PROCESSING;

        try {
            await this.onEvent(event);
        } catch (err) {
            this.state = AgentState.ERROR;
            this.log.error(`Error processing ${event.topic}: ${err}`);
            await this.emit("agent.error", {
                agentId: this.agentId,
                eventTopic: event.topic,
                error: err instanceof Error ? err.message : String(err),
            });
        } finally {
            if (this.state === AgentState.PROCESSING) this.state = AgentState.IDLE;
            this.processingLock = false;
        }
    }

    async onEvent(event: SwarmEvent): Promise<void> {
        const content = (event.payload.content as string) ?? (event.payload.message as string) ?? "";
        if (!content) return;

        const response = await this.think(content);
        if (response) {
            await this.emit(`agent.response.${this.config.name}`, {
                agentId: this.agentId,
                content: response.message.content,
                model: response.model,
                usage: response.usage,
                sourceEvent: event.eventId,
            });
        }
    }

    // ── Core Intelligence ───────────────────────────────────────

    async think(userInput: string): Promise<CompletionResponse | null> {
        if (!this.provider) {
            this.log.error("No provider bound");
            return null;
        }

        // Add user message
        this.messageHistory.push({ role: Role.USER, content: userInput });

        // Context pruning
        if (this.contextManager.needsPruning(this.messageHistory)) {
            this.log.debug("Context window full, pruning...");
            this.messageHistory = this.contextManager.prune(this.messageHistory);
        }

        const overrides: Record<string, unknown> = {};
        if (this.config.model) overrides.model = this.config.model;
        if (this.config.temperature !== undefined) overrides.temperature = this.config.temperature;

        let response: CompletionResponse;

        // If agent has tools, use the tool execution loop
        if (this.toolExecutor.registeredTools.length > 0) {
            const result = await this.toolExecutor.execute(this.provider, this.messageHistory);
            response = result.response;
            // Replace history with the full loop conversation
            this.messageHistory = result.messages;
        } else {
            response = await this.provider.complete(
                this.messageHistory,
                this.config.tools ?? undefined,
                overrides,
            );
            this.messageHistory.push(response.message);
        }

        // Track usage
        this.usageTracker.track(
            response.model,
            response.usage.promptTokens,
            response.usage.completionTokens,
        );

        // Persist to session
        if (this.sessionStore && this.currentSessionId) {
            this.sessionStore.addMessage(this.currentSessionId, { role: Role.USER, content: userInput });
            this.sessionStore.addMessage(this.currentSessionId, response.message);
        }

        this.log.debug(`Completed (${response.usage.totalTokens} tokens, ${this.usageTracker.summary.costFormatted} total)`);
        return response;
    }

    /**
     * Compact the conversation history using LLM summarization.
     */
    async compact(): Promise<void> {
        if (!this.provider) return;
        this.messageHistory = await this.contextManager.compact(this.messageHistory, this.provider);
    }

    /**
     * Reset the conversation (start a new session).
     */
    resetSession(): void {
        const systemMsgs = this.messageHistory.filter((m) => m.role === Role.SYSTEM);
        this.messageHistory = [...systemMsgs];

        if (this.sessionStore) {
            const session = this.sessionStore.create(this.agentId);
            this.currentSessionId = session.sessionId;
        }

        this.log.info("Session reset");
    }

    // ── Emission ────────────────────────────────────────────────

    async emit(
        topic: string,
        payload: Record<string, unknown>,
        priority: EventPriority = EventPriority.NORMAL,
    ): Promise<void> {
        await this.eventBus.publish(createEvent({ topic, payload, source: this.agentId, priority }));
    }

    // ── Streaming ──────────────────────────────────────────────

    /**
     * Stream a response token-by-token.
     * Yields each token as it arrives from the provider.
     */
    async *thinkStream(userInput: string): AsyncIterable<string> {
        if (!this.provider?.stream) {
            // Fallback to non-streaming
            const response = await this.think(userInput);
            if (response) yield response.message.content;
            return;
        }

        this.messageHistory.push({ role: Role.USER, content: userInput });

        if (this.contextManager.needsPruning(this.messageHistory)) {
            this.messageHistory = this.contextManager.prune(this.messageHistory);
        }

        const overrides: Record<string, unknown> = {};
        if (this.config.model) overrides.model = this.config.model;
        if (this.config.temperature !== undefined) overrides.temperature = this.config.temperature;

        let fullContent = "";
        for await (const token of this.provider.stream(this.messageHistory, undefined, overrides)) {
            fullContent += token;
            yield token;
        }

        // Add to history
        this.messageHistory.push({ role: Role.ASSISTANT, content: fullContent });

        // Persist
        if (this.sessionStore && this.currentSessionId) {
            this.sessionStore.addMessage(this.currentSessionId, { role: Role.USER, content: userInput });
            this.sessionStore.addMessage(this.currentSessionId, { role: Role.ASSISTANT, content: fullContent });
        }

        // Emit response event
        await this.emit(`agent.response.${this.config.name}`, {
            agentId: this.agentId,
            content: fullContent,
            streamed: true,
        });
    }

    // ── Introspection ───────────────────────────────────────────

    get history(): Message[] { return [...this.messageHistory]; }
    get contextUsage() { return this.contextManager.usage(this.messageHistory); }
    toString(): string { return `<Agent ${this.agentId} state=${this.state}>`; }
}
